### PR TITLE
Add ports data, to data_provider tool for Solaris vendor

### LINKED ISSUE #16010: Expand data provider for solaris ports
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.4 | Wazuh data provider | Agent | Packages/Sources | Solaris |

# Description

To have complete compatibility matrix support for this operating system it is necessary to add the capabilities to get this information from the operating system.

# DoD
- [ ] Spike where to get this information from.
- [ ] Development tasks.
- [ ] Add parser unit tests. (this test could be executed in linux)
- [ ] PR CI Green.
- [ ] Test e2e
- [ ] Create issue to update the documentation and work it.

### PR DESCRIPTION
**Pull request by tdrauncieal**
|Related issue|
|---|
|#16010|
## Description
It seeks to add functionality to acquire data related to the ports that are running in the Solaris terminal.
This closes #16010 

## Logs/Alerts example
An extract can be seen below running `testtool`:

## Tests

When proceed, this section should include new configuration parameters.

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Solaris
  - [x] Compiler GCC 5.5
  - [x] Compiler GCC 9.4
- [ ] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities
- [x] Added unit tests (for new features)

### CODE REVIEW

File: src/data_provider/src/ports/portSolarisWrapper.h [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+};
+
+class SolarisPortWrapper final : public IPortWrapper
+{
+        const mib2_tcpConnEntry_t* tcpEntry;
+        const mib2_tcp6ConnEntry_t* tcp6Entry;
+        const mib2_udpEntry_t* udpEntry;
+        const mib2_udp6Entry_t* udp6Entry;
+
+    public:
```
Comment (pereyra-m) [inline src/data_provider/src/ports/portSolarisWrapper.h:59]: Have you considered a C++ trick to avoid the four similar structs and their constructors? For example, `std::variant`

File: src/data_provider/src/sysInfoSolaris.cpp [hunk 2]
```diff
@@ -197,10 +200,82 @@ nlohmann::json SysInfo::getNetworks() const
...
+        PortSolarisHelper::mibGetItems(sd, items);
+
+        // Extract constant sizes - need do once only
+        PortSolarisHelper::mibGetConstants(items);
+
+        for (auto item : items)
+        {
+            if (item.group == MIB2_TCP)
+            {
+                for (tp = (mib2_tcpConnEntry_t*)item.val.get();
```
Comment (pereyra-m) [inline src/data_provider/src/sysInfoSolaris.cpp:229]: Here we have four times the same code, and we only change the structure we are using.
Have you considered a C++ trick to avoid this? For example, `std::variant`

**Review by pereyra-m — COMMENTED** [review]
Great job!

File: src/data_provider/src/ports/portSolarisHelper.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+int PortSolarisHelper::udp6EntrySize;
+
+void PortSolarisHelper::mibGetItems(int sd, std::vector<mib_item_t>& items)
+{
+    /*
+     * buf is an automatic for this function, so the
+     * compiler has complete control over its alignment;
+     * it is assumed this alignment is satisfactory for
+     * it to be casted to certain other struct pointers
+     * here, such as struct T_optmgmt_ack * .
```
Comment (MiguelazoDS) [inline src/data_provider/src/ports/portSolarisHelper.cpp:32]: Could you rephrase this? what is "alignment"?

File: src/data_provider/src/ports/portSolarisHelper.cpp [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        return (-1);
+    }
+
+    return (sd);
+}
+
+/* Extract constant sizes */
+void PortSolarisHelper::mibGetConstants(std::vector<mib_item_t>& items)
+{
+    /* 'for' loop 1: */
```
Comment (MiguelazoDS) [inline src/data_provider/src/ports/portSolarisHelper.cpp:162]: What's the meaning for these "for loop 1" comments?

File: src/data_provider/src/sysInfoSolaris.cpp [hunk 5]
```diff
@@ -197,10 +200,82 @@ nlohmann::json SysInfo::getNetworks() const
...
+
+    // Open /dev/arp and push into ioctl()
+    int sd = PortSolarisHelper::mibOpen();
+
+    if (sd != -1)
+    {
+        mib2_tcpConnEntry_t* tp;
+        mib2_tcp6ConnEntry_t* tp6;
+        mib2_udpEntry_t* ud;
+        mib2_udp6Entry_t* ud6;
```
Comment (MiguelazoDS) [inline src/data_provider/src/sysInfoSolaris.cpp:217]: Why not tcp, tcp6, udp, and udp6?

File: src/data_provider/src/sysInfoSolaris.cpp [hunk 6]
```diff
@@ -197,10 +200,82 @@ nlohmann::json SysInfo::getNetworks() const
...
 nlohmann::json SysInfo::getPorts() const
 {
-    return nlohmann::json();
+    nlohmann::json ports{};
+    std::vector<mib_item_t> items;
+
+    // Open /dev/arp and push into ioctl()
+    int sd = PortSolarisHelper::mibOpen();
+
+    if (sd != -1)
```
[the remark this answers is not in the document]
Comment (matias-braida) [reply src/data_provider/src/sysInfoSolaris.cpp:212]: If using int, check for "> 0" for successful case

File: src/data_provider/src/ports/portSolarisHelper.cpp [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    /*
+     * Each reply consists of a ctl part for one fixed structure
+     * or table, as defined in mib2.h.  The format is a T_OPTMGMT_ACK,
+     * containing an opthdr structure.  level/name identify the entry,
+     * len is the size of the data part of the message.
+     */
+    req = (struct opthdr*)&toa[1];
+    ctlbuf.maxlen = sizeof (buf);
+
+    while (1)
```
[the remark this answers is not in the document]
Comment (tdrauncieal) [reply src/data_provider/src/ports/portSolarisHelper.cpp:75]: I want to try but the code was get from a opensolaris's netstat source code.

File: src/data_provider/src/sysInfoSolaris.cpp [hunk 8]
```diff
@@ -197,10 +200,82 @@ nlohmann::json SysInfo::getNetworks() const
...
+        // Extract constant sizes - need do once only
+        PortSolarisHelper::mibGetConstants(items);
+
+        for (auto item : items)
+        {
+            if (item.group == MIB2_TCP)
+            {
+                for (tp = (mib2_tcpConnEntry_t*)item.val.get();
+                        (char*)tp < (char*)item.val.get() + item.length;
+                        tp = (mib2_tcpConnEntry_t*)((char*)tp + PortSolarisHelper::tcpConnEntrySize))
```
[the remark this answers is not in the document]
Comment (tdrauncieal) [reply src/data_provider/src/sysInfoSolaris.cpp:231]: Some reinterpret_cast will add.

File: src/data_provider/tests/sysInfoPortsSolaris/sysInfoSolarisPorts_test.cpp [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+    EXPECT_EQ(0, port.at("tx_queue").get<const int32_t>());
+    EXPECT_EQ(0, port.at("rx_queue").get<const int32_t>());
+    EXPECT_EQ(0, port.at("inode").get<const int32_t>());
+    EXPECT_EQ("2", port.at("state").get_ref<const std::string&>());
+    EXPECT_EQ(84, port.at("pid").get<const int32_t>());
+    EXPECT_EQ("", port.at("process").get_ref<const std::string&>());
+}
+
+/**
+ * @brief Test Not success
```
[the remark this answers is not in the document]
Comment (tdrauncieal) [reply src/data_provider/tests/sysInfoPortsSolaris/sysInfoSolarisPorts_test.cpp:203]: It's hard to reach!

File: src/data_provider/src/ports/portSolarisWrapper.h (hunk 1 above)
Comment (tdrauncieal) [reply src/data_provider/src/ports/portSolarisWrapper.h:59]: making use of this alternative makes the code excessively complex, which is originally intended to be C

File: src/data_provider/src/sysInfoSolaris.cpp (hunk 2 above)
Comment (tdrauncieal) [reply src/data_provider/src/sysInfoSolaris.cpp:229]: making use of this alternative makes the code excessively complex, which is originally intended to be C

File: src/data_provider/src/sysInfoSolaris.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tdrauncieal) [reply src/data_provider/src/sysInfoSolaris.cpp:212]: I don't think so, `sd` is a file identifier

File: src/data_provider/src/ports/portSolarisHelper.cpp [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+        items.push_back(item);
+    }
+}
+
+int PortSolarisHelper::mibOpen(void)
+{
+    int sd;
+
+    sd = open("/dev/arp", O_RDWR);
```
[the remark this answers is not in the document]
Comment (tdrauncieal) [reply src/data_provider/src/ports/portSolarisHelper.cpp:134]: this code was from OpenSolaris's netstat source code, but I test opened this for read-only.

File: src/data_provider/src/ports/portSolarisHelper.cpp (hunk 3 above)
Comment (tdrauncieal) [reply src/data_provider/src/ports/portSolarisHelper.cpp:32]: I leave this original comment, to accomplish

File: src/data_provider/src/sysInfoSolaris.cpp (hunk 5 above)
Comment (tdrauncieal) [reply src/data_provider/src/sysInfoSolaris.cpp:217]: sure!

File: src/data_provider/src/ports/portSolarisHelper.cpp (hunk 4 above)
Comment (tdrauncieal) [reply src/data_provider/src/ports/portSolarisHelper.cpp:162]: original comment will remove!

File: src/data_provider/src/sysInfoSolaris.cpp [hunk 11]
```diff
@@ -197,10 +200,82 @@ nlohmann::json SysInfo::getNetworks() const
...
+    // Open /dev/arp and push into ioctl()
+    int sd = PortSolarisHelper::mibOpen();
+
+    if (sd != -1)
+    {
+        mib2_tcpConnEntry_t* tp;
+        mib2_tcp6ConnEntry_t* tp6;
+        mib2_udpEntry_t* ud;
+        mib2_udp6Entry_t* ud6;
+
```
[the remark this answers is not in the document]
Comment (tdrauncieal) [reply src/data_provider/src/sysInfoSolaris.cpp:218]: these structures are part of the standard library in Solaris

File: src/data_provider/src/sysInfoSolaris.cpp [hunk 12]
```diff
@@ -197,10 +200,82 @@ nlohmann::json SysInfo::getNetworks() const
 
     return networks;
 }
+
 nlohmann::json SysInfo::getPorts() const
 {
-    return nlohmann::json();
+    nlohmann::json ports{};
+    std::vector<mib_item_t> items;
```
[the remark this answers is not in the document]
Comment (tdrauncieal) [reply src/data_provider/src/sysInfoSolaris.cpp:207]: these structures are part of the standard library in Solaris

File: src/data_provider/src/ports/portSolarisHelper.cpp [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+#include <memory>
+
+#include "portSolarisHelper.hpp"
+
+int PortSolarisHelper::tcpConnEntrySize;
+int PortSolarisHelper::tcp6ConnEntrySize;
+int PortSolarisHelper::udpEntrySize;
+int PortSolarisHelper::udp6EntrySize;
+
+void PortSolarisHelper::mibGetItems(int sd, std::vector<mib_item_t>& items)
```
Comment (matias-braida) [inline src/data_provider/src/ports/portSolarisHelper.cpp:25]: If "items" is not required to be contiguous in memory, there are better STL containers to use rather than std::vector. This is because every time you push_back an element into a std::vector, the whole vector is reallocated to a new contiguous heap memory area including the copy operation of all its items.
Maybe std::deque would be a better STL container type.

**Review by matias-braida — COMMENTED** [review]
GJ

File: src/data_provider/src/ports/portSolarisHelper.cpp (hunk 13 above)
Comment (tdrauncieal) [reply src/data_provider/src/ports/portSolarisHelper.cpp:25]: precisely in this case it is trying to emulate a vector C

Comment (matias-braida) [reply src/data_provider/src/ports/portSolarisHelper.cpp]: Where is it used or needed as C style vector?
I don't see it.

Comment (tdrauncieal) [reply src/data_provider/src/ports/portSolarisHelper.cpp]: My intention was not to stray too far from the original C code, but the change was made.

**Review by matias-braida — APPROVED** [review]

**Review by pereyra-m — APPROVED** [review]
